extensions/src/main/scala/ai/h2o/sparkling/extensions/rest/api/ImportFrameHandler.scala

```diff
@@ -41,28 +42,83 @@ class ImportFrameHandler extends Handler {
+    }
+    val domains = expandDomains(stringDomains, columnTypesAwareOfEmptyFrames)
+    ChunkUtils.finalizeFrame(request.key, rowsPerChunk, columnTypesAwareOfEmptyFrames, domains)
+    val frame = DKV.getGet[Frame](frameKey)
 
-    ChunkUtils.finalizeFrame(request.key, rowsPerChunk, columnTypes, domains)
+    convertColumnsWithTooManyCategoricalLevelsToStringColumns(frame, columnTypesAwareOfEmptyFrames, stringDomains)
 
-    // Update categorical indices for all chunks.
-    val frame = DKV.getGet[Frame](frameKey)
     val categoricalColumnIndices = for ((vec, idx) <- frame.vecs().zipWithIndex if vec.isCategorical) yield idx
+
+    // Update categorical indices for all chunks. (local chunk domains -> global domains)
     val updateCategoricalIndicesTask = new UpdateCategoricalIndicesTask(frameKey, categoricalColumnIndices)
     updateCategoricalIndicesTask.doAll(frame)
 
-    // Convert categorical columns with too many categories to T_STR
-    categoricalColumnIndices.foreach { catColIdx =>
-      val vector = frame.vec(catColIdx)
-      if (vector.cardinality() > Categorical.MAX_CATEGORICAL_COUNT) {
-        Log.warn(
-          s"Categorical column with index '$catColIdx' exceeded maximum number of categories. " +
+    // Convert unique categorical columns to T_STR
+    convertUniqueCategoricalColumnsToStringColumns(frame, categoricalColumnIndices)
+
+    request
+  }
+
+  private def convertColumnsWithTooManyCategoricalLevelsToStringColumns(
+      frame: Frame,
+      columnTypes: Array[Byte],
+      stringDomains: Array[Array[String]]) = {
+    val indicesOfChanges = indicesOfChangedColumnsFromCategoricalToString(columnTypes, stringDomains)
+    if (indicesOfChanges.nonEmpty) {
+      indicesOfChanges.foreach { idx =>
+        Log.info(
+          s"A categorical column '${frame.names()(idx)}' exceeded maximum number of categories. " +
             "Converting it to a column of strings ...")
-        frame.replace(catColIdx, vector.toStringVec())
+      }
+      val vectorsToBeConvertedToString = indicesOfChanges.map(frame.vec(_))
+      val domainIndices = for ((domain, idx) <- stringDomains.zipWithIndex if domain == null) yield idx
+      val convertCategoricalToStringColumnsTask = new ConvertCategoricalToStringColumnsTask(frame._key, domainIndices)
+      convertCategoricalToStringColumnsTask.doAll(Vec.T_STR, vectorsToBeConvertedToString: _*)
+      val newVectors = AppendableVec.closeAll(convertCategoricalToStringColumnsTask.appendables())
+      for ((newVector, index) <- newVectors.zip(indicesOfChanges)) {
+        val oldVector = frame.replace(index, newVector)
+        oldVector.remove()
       }
     }
+  }
 
-    request
+  private def convertUniqueCategoricalColumnsToStringColumns(frame: Frame, categoricalColumnIndices: Array[Int]) = {
+    categoricalColumnIndices.foreach { idx =>
+      val vector = frame.vec(idx)
+      val ratio = vector.cardinality() / vector.length().asInstanceOf[Double]
+      if (ratio > 0.95) {
```

Comment:
The heuristic logic utilized by H2O parsers (`PreviewParseWriter`):
```
  @Override public void addStrCol(int colIdx, BufferedString str) {
    if(colIdx < _ncols) {
      // Check for time
      if (ParseTime.isTime(str)) {
        ++_ndates[colIdx];
        return;
      }

      //Check for UUID
      if(ParseUUID.isUUID(str)) {
        ++_nUUID[colIdx];
        return;
      }

      //Add string to domains list for later determining string, NA, or categorical
      ++_nstrings[colIdx];
      _domains[colIdx].put(str.toString(),"");

      if (_nlines < MAX_PREVIEW_LINES)
        _data[_nlines][colIdx] = str.toString();
    }
  }
```

```
      // UUID
      if (_nUUID[i] > _ndates[i]
              && _nUUID[i] > (_nnums[i] + _nzeros[i])
              && (_nUUID[i] > _nstrings[i] || _domains[i].size() <= 1)) {
        types[i] = Vec.T_UUID;
        continue;
      }

      // Strings, almost no dups
      if (_nstrings[i] > _ndates[i]
              && _nstrings[i] > _nUUID[i]
              && _nstrings[i] > (_nnums[i] + _nzeros[i])
              && _domains[i].size() >= 0.95 * _nstrings[i]) {
        types[i] = Vec.T_STR;
        continue;
      }

      // categorical or string?
      // categorical with 0s for NAs
      if(_nzeros[i] > 0
              && ((_nzeros[i] + _nstrings[i]) >= nonemptyLines) //just strings and zeros for NA (thus no empty lines)
              && (_domains[i].size() <= 0.95 * _nstrings[i]) ) { // not all unique strings
        types[i] = Vec.T_CAT;
        continue;
      }
      // categorical mixed with numbers
      if(_nstrings[i] >= (_nnums[i]+_nzeros[i]) // mostly strings
              && (_domains[i].size() <= 0.95 * _nstrings[i]) ) { // but not all unique
        types[i] = Vec.T_CAT;
        continue;
      }
```

I'm not sure whether it's necessary to cover also `Vec.T_TIME` and `Vec.T_UUID`. `Vec.T_TIME` is usually converted from Spark's `DateType` or `TimestampType`.

Reply:
I've tried to use `PreviewParseWriter` where Vec.T_TIME and Vec.T_UUID are unsupported.